File: server/core/predict.go
package core

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strconv"
	"strings"

	ort "github.com/yalue/onnxruntime_go"
)

// smartData is a more flexible struct that can handle both NVMe and SATA data.
type smartData struct {
	Device struct {
		Type     string `json:"type"`
		Protocol string `json:"protocol"`
	} `json:"device"`
	SmartStatus struct {
		Passed bool `json:"passed"`
	} `json:"smart_status"`
	AtaSmartAttributes struct {
		Table []struct {
			ID    int    `json:"id"`
			Name  string `json:"name"`
			Value int    `json:"value"`
			Raw   struct {
				Value int `json:"value"`
			} `json:"raw"`
		} `json:"table"`
	} `json:"ata_smart_attributes"`
	NvmeSmartHealthInformationLog struct {
		Temperature      int `json:"temperature"`
		PercentageUsed   int `json:"percentage_used"`
		DataUnitsWritten int `json:"data_units_written"`
		PowerOnHours     int `json:"power_on_hours"`
	} `json:"nvme_smart_health_information_log"`
}

// SmartAttribute holds the raw and normalized values for a S.M.A.R.T. attribute
type SmartAttribute struct {
	Name  string `json:"name"`
	Value int    `json:"value"`
}

// PredictionResult is what we'll send to the frontend
type PredictionResult struct {
	PredictedStatus    string                    `json:"predictedStatus"`
	FailureProbability float32                   `json:"failureProbability"`
	SmartStatus        string                    `json:"smartStatus"`
	SmartAttributes    map[string]SmartAttribute `json:"smartAttributes"`
}

func PredictDriveHealth(devicePath string) (*PredictionResult, error) {
	cmd := exec.Command("smartctl", "-a", "-j", devicePath)
	out, err := cmd.Output()
	if err != nil {
		// If smartctl fails, it could be a USB drive or a device that doesn't support S.M.A.R.T.
		return &PredictionResult{
			PredictedStatus: "N/A",
			SmartStatus:     "Not available",
			SmartAttributes: make(map[string]SmartAttribute),
		}, nil
	}

	var data smartData
	if err := json.Unmarshal(out, &data); err != nil {
		return nil, fmt.Errorf("failed to parse S.M.A.R.T. data: %w", err)
	}

	result := &PredictionResult{
		PredictedStatus:    "Healthy", // Default to Healthy
		FailureProbability: 0,
		SmartAttributes:    make(map[string]SmartAttribute),
	}

	if data.SmartStatus.Passed {
		result.SmartStatus = "Passed"
	} else {
		result.SmartStatus = "Failing"
		result.PredictedStatus = "At Risk"
	}

	// Handle different drive types
	if data.Device.Type == "nvme" {
		result.SmartAttributes["Temperature"] = SmartAttribute{Name: "Temperature", Value: data.NvmeSmartHealthInformationLog.Temperature}
		result.SmartAttributes["Percentage Used"] = SmartAttribute{Name: "Percentage Used", Value: data.NvmeSmartHealthInformationLog.PercentageUsed}
		result.SmartAttributes["Data Units Written"] = SmartAttribute{Name: "Data Units Written", Value: data.NvmeSmartHealthInformationLog.DataUnitsWritten}
		result.SmartAttributes["Power On Hours"] = SmartAttribute{Name: "Power On Hours", Value: data.NvmeSmartHealthInformationLog.PowerOnHours}
	} else if data.Device.Type == "sat" { // SATA drive
		// For SATA drives, we can use the ONNX model
		for _, attr := range data.AtaSmartAttributes.Table {
			switch attr.ID {
			case 5, 9, 177, 194, 241:
				result.SmartAttributes[attr.Name] = SmartAttribute{Name: attr.Name, Value: attr.Raw.Value}
			}
		}
		runSataPrediction(result, &data)
	} else {
		result.SmartStatus = "Not supported"
	}

	return result, nil
}

// runSataPrediction runs the ONNX model for SATA drives.
func runSataPrediction(result *PredictionResult, data *smartData) {
	// This block will attempt to run the prediction, but will not return an error if it fails.
	// It will log the error and the function will return the result with just the S.M.A.R.T. data.
	defer func() {
		if r := recover(); r != nil {
			log.Printf("Recovered from panic in runSataPrediction: %v", r)
		}
	}()

	featureMapFile, err := os.ReadFile("../model/feature_map.json")
	if err != nil {
		log.Printf("Warning: could not load feature_map.json: %v", err)
		return
	}
	var featureMap map[string]string
	if err := json.Unmarshal(featureMapFile, &featureMap); err != nil {
		log.Printf("Warning: could not parse feature_map.json: %v", err)
		return
	}

	featureNames := make([]string, len(featureMap))
	for key, value := range featureMap {
		index, err := strconv.Atoi(strings.TrimPrefix(value, "f"))
		if err != nil {
			log.Printf("Warning: could not parse feature index from %s", value)
			continue
		}
		if index < len(featureNames) {
			featureNames[index] = key
		}
	}

	inputTensor := make([]float32, len(featureNames))
	smartValues := make(map[string]int)
	for _, attr := range data.AtaSmartAttributes.Table {
		keyRaw := fmt.Sprintf("smart_%d_raw", attr.ID)
		keyNorm := fmt.Sprintf("smart_%d_normalized", attr.ID)
		smartValues[keyRaw] = attr.Raw.Value
		smartValues[keyNorm] = attr.Value
	}

	for i, feature := range featureNames {
		if val, ok := smartValues[feature]; ok {
			inputTensor[i] = float32(val)
		} else {
			inputTensor[i] = 0 // Handle missing values
		}
	}

	inputShape := ort.NewShape(1, int64(len(featureNames)))
	input, err := ort.NewTensor(inputShape, inputTensor)
	if err != nil {
		log.Printf("Warning: failed to create ONNX input tensor: %v", err)
		return
	}
	defer input.Destroy()

	outputShape := ort.NewShape(1, 2)
	output, err := ort.NewEmptyTensor[float32](outputShape)
	if err != nil {
		log.Printf("Warning: failed to create ONNX output tensor: %v", err)
		return
	}
	defer output.Destroy()

	model, err := ort.NewSession[float32]("../model/drive_failure_model.onnx",
		[]string{"float_input"}, []string{"output_probability"},
		[]*ort.Tensor[float32]{input}, []*ort.Tensor[float32]{output})
	if err != nil {
		log.Printf("Warning: failed to create ONNX session: %v", err)
		return
	}
	defer model.Destroy()

	err = model.Run()
	if err != nil {
		log.Printf("Warning: failed to run model prediction: %v", err)
		return
	}

	probabilities := output.GetData()
	failureProbability := probabilities[1]

	result.FailureProbability = failureProbability
	if failureProbability > 0.5 {
		result.PredictedStatus = "At Risk"
	} else {
		result.PredictedStatus = "Healthy"
	}
}

File: server/src/core/predict.rs
// Port of server-go/core/predict.go
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::process::Command;

use super::drives::log_line;

/// Raw and normalized values for a S.M.A.R.T. attribute
#[derive(Debug, Clone, Serialize)]
pub struct SmartAttribute {
    pub name: String,
    pub value: i64,
}

/// What we send to the frontend
#[derive(Debug, Clone, Serialize)]
pub struct PredictionResult {
    #[serde(rename = "predictedStatus")]
    pub predicted_status: String,
    #[serde(rename = "failureProbability")]
    pub failure_probability: f32,
    #[serde(rename = "smartStatus")]
    pub smart_status: String,
    #[serde(rename = "smartAttributes")]
    pub smart_attributes: HashMap<String, SmartAttribute>,
}

// Wrapper to parse smartctl's ata_smart_attributes table with raw.value.
#[derive(Debug, Deserialize)]
struct AtaAttributeRaw {
    #[serde(default)]
    id: i64,
    #[serde(default)]
    name: String,
    #[serde(default)]
    value: i64,
    #[serde(default)]
    raw: AtaRawValue,
}

#[derive(Debug, Default, Deserialize)]
struct AtaRawValue {
    #[serde(default)]
    value: i64,
}

pub fn predict_drive_health(device_path: &str) -> Result<PredictionResult, String> {
    let out = Command::new("smartctl")
        .args(["-a", "-j", device_path])
        .output()
        .map_err(|e| e.to_string());

    let out = match out {
        Ok(o) if o.status.success() => o,
        _ => {
            // If smartctl fails, it could be a USB drive or a device that
            // doesn't support S.M.A.R.T.
            return Ok(PredictionResult {
                predicted_status: "N/A".to_string(),
                failure_probability: 0.0,
                smart_status: "Not available".to_string(),
                smart_attributes: HashMap::new(),
            });
        }
    };

    let json: serde_json::Value = serde_json::from_slice(&out.stdout)
        .map_err(|e| format!("failed to parse S.M.A.R.T. data: {e}"))?;

    let mut result = PredictionResult {
        predicted_status: "Healthy".to_string(), // Default to Healthy
        failure_probability: 0.0,
        smart_status: String::new(),
        smart_attributes: HashMap::new(),
    };

    let passed = json
        .pointer("/smart_status/passed")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);
    if passed {
        result.smart_status = "Passed".to_string();
    } else {
        result.smart_status = "Failing".to_string();
        result.predicted_status = "At Risk".to_string();
    }

    let device_type = json
        .pointer("/device/type")
        .and_then(|v| v.as_str())
        .unwrap_or("");

    match device_type {
        "nvme" => {
            let log = &json["nvme_smart_health_information_log"];
            let get = |key: &str| log.get(key).and_then(|v| v.as_i64()).unwrap_or(0);
            for (name, key) in [
                ("Temperature", "temperature"),
                ("Percentage Used", "percentage_used"),
                ("Data Units Written", "data_units_written"),
                ("Power On Hours", "power_on_hours"),
            ] {
                result.smart_attributes.insert(
                    name.to_string(),
                    SmartAttribute {
                        name: name.to_string(),
                        value: get(key),
                    },
                );
            }
        }
        "sat" => {
            // SATA drive: we can use the ONNX model
            let table: Vec<AtaAttributeRaw> = serde_json::from_value(
                json.pointer("/ata_smart_attributes/table")
                    .cloned()
                    .unwrap_or(serde_json::json!([])),
            )
            .unwrap_or_default();
            for attr in &table {
                if matches!(attr.id, 5 | 9 | 177 | 194 | 241) {
                    result.smart_attributes.insert(
                        attr.name.clone(),
                        SmartAttribute {
                            name: attr.name.clone(),
                            value: attr.raw.value,
                        },
                    );
                }
            }
            run_sata_prediction(&mut result, &table);
        }
        _ => {
            result.smart_status = "Not supported".to_string();
        }
    }

    Ok(result)
}

/// Runs the ONNX model for SATA drives.
/// Attempts the prediction but never fails hard: on any error it logs a
/// warning and returns the result with just the S.M.A.R.T. data.
fn run_sata_prediction(result: &mut PredictionResult, table: &[AtaAttributeRaw]) {
    if let Err(e) = try_run_sata_prediction(result, table) {
        log_line(&format!("Warning: SATA prediction skipped: {e}"));
    }
}

fn try_run_sata_prediction(
    result: &mut PredictionResult,
    table: &[AtaAttributeRaw],
) -> Result<(), String> {
    let feature_map_file = std::fs::read_to_string("../model/feature_map.json")
        .map_err(|e| format!("could not load feature_map.json: {e}"))?;
    let feature_map: HashMap<String, String> = serde_json::from_str(&feature_map_file)
        .map_err(|e| format!("could not parse feature_map.json: {e}"))?;

    let mut feature_names: Vec<String> = vec![String::new(); feature_map.len()];
    for (key, value) in &feature_map {
        let index: usize = match value
            .trim_start_matches('f')
            .parse()
        {
            Ok(i) => i,
            Err(_) => {
                log_line(&format!(
                    "Warning: could not parse feature index from {value}"
                ));
                continue;
            }
        };
        if index < feature_names.len() {
            feature_names[index] = key.clone();
        }
    }

    let mut smart_values: HashMap<String, i64> = HashMap::new();
    for attr in table {
        smart_values.insert(format!("smart_{}_raw", attr.id), attr.raw.value);
        smart_values.insert(format!("smart_{}_normalized", attr.id), attr.value);
    }

    let input_tensor: Vec<f32> = feature_names
        .iter()
        .map(|feature| smart_values.get(feature).copied().unwrap_or(0) as f32)
        .collect();

    let n = input_tensor.len();
    let input = ort::value::Tensor::from_array(([1usize, n], input_tensor))
        .map_err(|e| format!("failed to create ONNX input tensor: {e}"))?;

    let mut session = ort::session::Session::builder()
        .and_then(|mut b| b.commit_from_file("../model/drive_failure_model.onnx"))
        .map_err(|e| format!("failed to create ONNX session: {e}"))?;

    let outputs = session
        .run(ort::inputs![input])
        .map_err(|e| format!("failed to run model prediction: {e}"))?;

    let (_, probabilities) = outputs["output_probability"]
        .try_extract_tensor::<f32>()
        .map_err(|e| format!("failed to read ONNX output tensor: {e}"))?;

    let failure_probability = probabilities.get(1).copied().unwrap_or(0.0);

    result.failure_probability = failure_probability;
    if failure_probability > 0.5 {
        result.predicted_status = "At Risk".to_string();
    } else {
        result.predicted_status = "Healthy".to_string();
    }
    Ok(())
}
